contracts/multisig-prover/src/command.rs

```diff
@@ -0,0 +1,74 @@
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::HexBinary;
+use sha3::{Digest, Keccak256};
+
+use axelar_wasm_std::hash::Hash;
+use multisig::worker_set::WorkerSet;
+use router_api::{CrossChainId, Message};
+
+use crate::{
+    encoding::{abi2, Encoder},
+    types::BatchId,
+};
+
+#[cw_serde]
+pub enum Payload {
+    Messages(Vec<Message>),
+    WorkerSet(WorkerSet),
+}
+
+impl Payload {
+    // id returns the unique identifier for the payload, which can be either
+    // - the hash of comma separated sorted message ids
+    // - the hash of the worker set
+    pub fn id(&self) -> BatchId {
+        match &self {
+            Payload::Messages(msgs) => {
+                let mut message_ids = msgs
+                    .iter()
+                    .map(|msg| msg.cc_id.to_string())
+                    .collect::<Vec<_>>();
+                message_ids.sort();
+
+                Keccak256::digest(message_ids.join(",")).as_slice().into()
+            }
+            Payload::WorkerSet(worker_set) => worker_set.hash().into(),
+        }
+    }
+}
+
+#[cw_serde]
+pub struct Command {
+    pub id: BatchId,
+    pub payload: Payload,
+}
+
+impl Command {
+    pub fn new(payload: Payload) -> Self {
+        Command {
+            id: payload.id(),
+            payload,
+        }
+    }
+
+    pub fn msg_digest(
+        &self,
+        encoder: Encoder,
+        domain_separator: &Hash,
+        curr_worker_set: &WorkerSet,
+    ) -> HexBinary {
+        match encoder {
+            Encoder::Abi => {
+                abi2::message_hash_to_sign(domain_separator, curr_worker_set, &self.payload)
+            }
+            Encoder::Bcs => todo!(),
+        }
+    }
+
+    pub fn message_ids(&self) -> Vec<CrossChainId> {
+        match &self.payload {
+            Payload::Messages(msgs) => msgs.iter().map(|msg| msg.cc_id.clone()).collect(),
+            Payload::WorkerSet(_) => vec![],
+        }
+    }
```

Comment:
not saying I like that, following the precedence https://github.com/axelarnetwork/axelar-amplifier/blob/main/contracts/multisig-prover/src/types.rs#L90, it returns empty vector if the batch contains update worker set. I can remove the function in this pr, it will be used GetProofResponse query. will improve in the follow up pr.